src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -404,5 +407,230 @@ public static unsafe void ClearWithReferences(ref IntPtr ip, nuint pointerSizeLe
+                last = ref Unsafe.Add(ref Unsafe.Add(ref first, numBytes), -Vector128<byte>.Count);
+                do
+                {
+                    // Load in values from beginning and end of the array.
+                    Vector128<byte> tempFirst = Unsafe.As<byte, Vector128<byte>>(ref first);
+                    Vector128<byte> tempLast = Unsafe.As<byte, Vector128<byte>>(ref last);
+
+                    // Shuffle to reverse each vector:
+                    //     +---------------+
+                    //     | A | B | C | D |
+                    //     +---------------+
+                    //          --->
+                    //     +---------------+
+                    //     | D | C | B | A |
+                    //     +---------------+
+                    tempFirst = Ssse3.Shuffle(tempFirst, ReverseMask);
+                    tempLast = Ssse3.Shuffle(tempLast, ReverseMask);
+
+                    // Store the reversed vectors
+                    Unsafe.As<byte, Vector128<byte>>(ref first) = tempLast;
+                    Unsafe.As<byte, Vector128<byte>>(ref last) = tempFirst;
+
+                    // Adjust the references to point to the next vector
+                    first = ref Unsafe.Add(ref first, Vector128<byte>.Count);
+                    last = ref Unsafe.Add(ref last, -Vector128<byte>.Count);
+                    numBytesWritten += Vector128<byte>.Count * 2;
+                } while ((numBytes - numBytesWritten) >= Vector128<byte>.Count * 2);
+            }
+
+            // Store any remaining values one-by-one
+            last = ref Unsafe.Add(ref first, (numBytes - numBytesWritten) - sizeof(int));
+            ref int firstInt = ref Unsafe.As<byte, int>(ref first);
+            ref int lastInt = ref Unsafe.As<byte, int>(ref last);
+            while (numBytes - numBytesWritten > sizeof(int))
+            {
+                (lastInt, firstInt) = (firstInt, lastInt);
+                firstInt = ref Unsafe.Add(ref firstInt, 1);
+                lastInt = ref Unsafe.Add(ref lastInt, -1);
+                numBytesWritten += sizeof(int) * 2;
+            }
+        }
+
+        public static void Reverse(ref long buf, nuint length)
+        {
+            nint numBytes = (int)length * sizeof(long);
+            int numBytesWritten = 0;
+            ref byte first = ref Unsafe.As<long, byte>(ref buf);
+            ref byte last = ref Unsafe.Add(ref Unsafe.Add(ref first, numBytes), -sizeof(long));
+            if (Avx2.IsSupported && Vector256<byte>.Count * 2 <= numBytes)
+            {
+                Vector256<byte> ReverseMask = Vector256.Create(
+                    (byte)8, 9, 10, 11, 12, 13, 14, 15, 0, 1, 2, 3, 4, 5, 6, 7, // first 128-bit lane
+                     8, 9, 10, 11, 12, 13, 14, 15, 0, 1, 2, 3, 4, 5, 6, 7); // second 128-bit lane
+                last = ref Unsafe.Add(ref Unsafe.Add(ref first, numBytes), -Vector256<byte>.Count);
```

Comment:
Semantically, using [Subtract](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.subtract?view=net-6.0) would be better.
```suggestion
                last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, numBytes), Vector256<byte>.Count);
```

Reply:
Updated.